lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,795 @@
+        if let InsertState::Losing { count, timestamp } = self.insert_state {
+            // Note: already includes OVERHEAD/DATA_LOSS_LEN
+            let room = self.free_space();
+            let required = space_required.unwrap_or(0);
+            if room >= required {
+                // We can recover!
+                self.write_header(
+                    self.our_task_id,
+                    timestamp,
+                    DATA_LOSS_LEN as u16,
+                );
+
+                // CBOR-format loss record:
+                // a1  # map(1)
+                //    # key
+                //   64 6c 6f 73 74  # text("lost")
+                //    # value
+                //   1a 00 00 00 00
+
+                for byte in [0xA1, 0x64, 0x6C, 0x6F, 0x73, 0x74, 0x1A] {
+                    self.storage.push_back(byte).unwrap_lite();
+                }
+                for byte in u32::from(count).to_be_bytes() {
+                    self.storage.push_back(byte).unwrap_lite();
+                }
+
+                self.insert_state = InsertState::Collecting;
+                true
+            } else {
+                // Recovery is not currently possible (not enough room).
+                false
+            }
+        } else {
+            // We're good.
+            true
+        }
+    }
+}
+
+/// Represents a record in the queue, with its header parsed.
+///
+/// The body of the record is given as a _pair_ of slices because it's not
+/// necessarily contiguous in the internal deque.
+#[derive(Copy, Clone, Debug)]
+pub struct Record<'s> {
+    /// The event's ENA.
+    pub ena: u64,
+    /// The sender's task ID.
+    pub tid: u16,
+    /// The timestamp from the report.
+    pub timestamp: u64,
+    /// The contents of the report.
+    pub slices: (&'s [u8], &'s [u8]),
+}
+
+impl Record<'_> {
+    /// Iterates over the body bytes, hiding the fact that they be
+    /// discontiguous.
+    pub fn body_bytes(&self) -> impl Iterator<Item = u8> + '_ {
+        self.slices
+            .0
+            .iter()
+            .chain(self.slices.1.iter())
+            .copied()
+    }
+}
+
+/// Utility function for pulling a fixed number of bytes from the front of a
+/// pair of slices.
+fn take_array<'a, const N: usize>(
+    slices: &mut (&'a [u8], &'a [u8]),
+) -> Option<[u8; N]> {
+    let (first, second) = take_slice(slices, N)?;
+
+    let mut result = [0; N];
+    result[..first.len()].copy_from_slice(first);
+    result[first.len()..].copy_from_slice(second);
+
+    Some(result)
+}
+
+/// Utility function for pulling a variable number of bytes from the front of a
+/// pair of slices.
+fn take_slice<'a>(
+    slices: &mut (&'a [u8], &'a [u8]),
+    n: usize,
+) -> Option<(&'a [u8], &'a [u8])> {
+    let first_n = usize::min(slices.0.len(), n);
+    let (first, rest) = slices.0.split_at(first_n);
+    slices.0 = rest;
+
+    let rest_n = n - first_n;
+    if slices.1.len() < rest_n {
+        return None;
+    }
```

Comment:
So, I'm not sure if this is an issue in practice, but it occurs to me that, in the case where `slices.1.len() < rest_n`, we will have mutated `slices.0` but return `None`, which potentially leaves the two slices in a weird-ish state. Should we instead consider performing the check that there are enough bytes to take _before_ we mutate either slice, as in:
```suggestion
    let rest_n = n - first_n;
    if slices.1.len() < rest_n {
        return None;
    }
    
    let (first, rest) = slices.0.split_at(first_n);
    slices.0 = rest;
```

Reply:
I think this is a great point -- while it doesn't make a high level difference in the behavior right now, I think it's important to get the semantics of the utility functions right, or they'll break something subtle during maintenance later.

Reply:
Yup, that was my thinking as well, it just felt like a sharp edge that we weren't *currently* grabbing but might as well sand down in case we reach for it later. Thanks for fixing it!